### PR TITLE
Activate client certs in local-cluster-up.sh for RBAC

### PR DESCRIPTION
**Pull request by sttts**
- set up client certs for each component in local-cluster-up.sh

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36707)

### CODE REVIEW

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -362,13 +390,23 @@ function start_apiserver {
     # Ensure CERT_DIR is created for auto-generated crt/key and kubeconfig
     sudo mkdir -p "${CERT_DIR}"
 
+    # Create client ca
+    sudo /bin/bash -e <<EOF
+    rm -f "${CERT_DIR}/client-ca.crt" "${CERT_DIR}/client-ca.key"
+    openssl req -x509 -sha256 -new -nodes -days 365 -newkey rsa:2048 -keyout "${CERT_DIR}/client-ca.key" -out "${CERT_DIR}/client-ca.crt" -subj "/C=xx/ST=x/L=x/O=x/OU=x/CN=ca/emailAddress=x/"
+    echo '{"signing":{"default":{"expiry":"43800h","usages":["signing","key encipherment","client auth"]}}}' > "${CERT_DIR}/client-ca-config.json"
+EOF
+    create_client_certkey kubelet
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: Nodes should have a name like `system:node:<node-name>` and  the `system:nodes` group.

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -362,13 +390,23 @@ function start_apiserver {
...
+    create_client_certkey kube-proxy
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: I'd like to see this as another `system:kube-proxy` style user.   We've reserved the `system:`  prefix throughout the authn/authz layers for "the kube cluster relies on this stuff, don't muck with it unless you really know what you're doing" resources.

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -362,13 +390,24 @@ function start_apiserver {
...
+    create_client_certkey controller
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: This is a controller-bootstrapper identity.  We expect controllers to run as service accounts and only bootstrapping will require special powers.  It should have a `system:masters` group.

File: hack/local-up-cluster.sh [hunk 4]
```diff
@@ -362,13 +390,24 @@ function start_apiserver {
...
+    create_client_certkey scheduler
+    create_client_certkey admin
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: name should be `system:admin` and it should have a `system:cluster-admins` group.

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (sttts) [inline hack/local-up-cluster.sh]: kubelet => system:node:${HOSTNAME_OVERRIDE}
kube-proxy => ?

File: hack/local-up-cluster.sh [hunk 5]
```diff
@@ -316,6 +316,38 @@ function set_service_accounts {
     fi
 }
 
+function create_client_certkey {
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: I'd like to be allowed to pass a CA through to here so I can create one for front-proxy authentication easily.

File: hack/local-up-cluster.sh [hunk 6]
```diff
@@ -362,13 +390,24 @@ function start_apiserver {
...
+    echo '{"signing":{"default":{"expiry":"43800h","usages":["signing","key encipherment","client auth"]}}}' > "${CERT_DIR}/client-ca-config.json"
+EOF
+    create_client_certkey kubelet
+    create_client_certkey kube-proxy
+    create_client_certkey controller
+    create_client_certkey scheduler
+    create_client_certkey admin
 
     APISERVER_LOG=/tmp/kube-apiserver.log
     sudo -E "${GO_OUT}/hyperkube" apiserver ${anytoken_arg} ${authorizer_arg} ${priv_arg} ${runtime_config}\
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: I'd like to see us stop passing cert_dir and pass explicit CA and serving certs.

File: hack/local-up-cluster.sh [hunk 7]
```diff
@@ -316,6 +316,38 @@ function set_service_accounts {
...
+    echo '{"CN":"$1","hosts":[""],"key":{"algo":"rsa","size":2048}}' | docker run -i  --entrypoint /bin/bash -v "${CERT_DIR}:/certs" -w /certs cfssl/cfssl:latest -ec "cfssl gencert -ca=client-ca.crt -ca-key=client-ca.key -config=client-ca-config.json - | cfssljson -bare client-$1"
+    sudo /bin/bash -e <<EOF
+    mv "${CERT_DIR}/client-$1-key.pem" "${CERT_DIR}/client-$1.key"
+    mv "${CERT_DIR}/client-$1.pem" "${CERT_DIR}/client-$1.crt"
+    rm -f "${CERT_DIR}/client-$1.csr"
+EOF
+}
+
+function write_client_kubeconfig {
+    cat <<EOF | sudo tee "${CERT_DIR}"/kubeconfig-$1 > /dev/null
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: I'd like to see the name as `$1.kubeconfig`.

**Comment by deads2k** [comment]
This is a great step forward.

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (deads2k) [inline hack/local-up-cluster.sh]: `system:proxy`? I don't think the proxy permissions vary by node.

Comment (liggitt) [inline hack/local-up-cluster.sh]: we have a lot of proxies... `system:kube-proxy` might be better. We'll eventually want a client cert for the master proxy -> pods/services/nodes, and we used `system:master-proxy` for that (which grants no API powers, just identifies the api proxy to backends)

**Comment by deads2k** [comment]
lgtm, go ahead squash down.

File: hack/local-up-cluster.sh [hunk 8]
```diff
@@ -362,13 +401,27 @@ function start_apiserver {
...
 
+    # Create client ca
+    sudo /bin/bash -e <<EOF
+    rm -f "${CERT_DIR}/client-ca.crt" "${CERT_DIR}/client-ca.key"
+    openssl req -x509 -sha256 -new -nodes -days 365 -newkey rsa:2048 -keyout "${CERT_DIR}/client-ca.key" -out "${CERT_DIR}/client-ca.crt" -subj "/C=xx/ST=x/L=x/O=x/OU=x/CN=ca/emailAddress=x/"
+    echo '{"signing":{"default":{"expiry":"43800h","usages":["signing","key encipherment","client auth"]}}}' > "${CERT_DIR}/client-ca-config.json"
+EOF
+
+    # Create client certs signed with client-ca, given id, given CN and a number of groups
+    # NOTE: system:masters will be removed in the future
```
Comment (nikhiljindal) [inline hack/local-up-cluster.sh]: Why?
(Just curious)

Comment (sttts) [inline hack/local-up-cluster.sh]: @deads2k ^^

Comment (deads2k) [inline hack/local-up-cluster.sh]: > Why?
> (Just curious)

Controllers should start running with their own identity.  Part of that will involved a bootstrapping identity with sufficient power to create services accounts and run the SA token secret generator, but it should not require full access.

RBAC will authorize based on specific permissions and non-RBAC authorizers will be able to authorize based on username, but neither need to grant full powers.

**Comment by deads2k** [comment]
@k8s-bot unit test this

**Comment by deads2k** [comment]
@k8s-bot unit test this: issue #34953

**Pull request merged**